Windows Node errors while trying to load Pyodide

## 🐛 Bug

When I try to load pyodide in a Windows Node environment, I get the following error:

```bash
Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'd:'
    at new NodeError (node:internal/errors:387:5)
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/resolve:1075:11)
    at defaultResolve (node:internal/modules/esm/resolve:1155:3)
    at nextResolve (node:internal/modules/esm/loader:173:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:852:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:439:18)
    at ESMLoader.import (node:internal/modules/esm/loader:536:22)
    at importModuleDynamically (node:internal/modules/cjs/loader:1059:29)
    at importModuleDynamicallyWrapper (node:internal/vm/module:438:21)
    at importModuleDynamically (node:vm:389:46) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}
```

I have a local patch to fix the above issue in `nodeLoadScript()` that I will submit a PR for tomorrow. However, I then run into the following issue:

```bash
Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\Projects\pyodide\dist\pyodide.j\pyodide.asm.js' imported from D:\Projects\pyodide\dist\pyodide.js
    at new NodeError (node:internal/errors:387:5)
    at finalizeResolution (node:internal/modules/esm/resolve:404:11)
    at moduleResolve (node:internal/modules/esm/resolve:965:10)
    at defaultResolve (node:internal/modules/esm/resolve:1173:11)
    at nextResolve (node:internal/modules/esm/loader:173:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:852:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:439:18)
    at ESMLoader.import (node:internal/modules/esm/loader:536:22)
    at importModuleDynamically (node:internal/modules/cjs/loader:1059:29)
    at importModuleDynamicallyWrapper (node:internal/vm/module:438:21) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

I suspect that `calculateIndexURL()` is not properly extracting the location of the `pyodide.asm.js` file. I can submit a patch for this tomorrow as well unless y'all already have a good idea of how to fix it.

### To Reproduce

```
import { loadPyodide } from "pyodide/pyodide.js";

(async () => {
  const pyodide = await loadPyodide();
  await pyodide.runPythonAsync('print("Hello World!")');
})();
```

### Expected behavior

Windows Node should be able to load pyodide properly.

### Environment

- Pyodide Version<!-- (e.g. 1.8.1) -->: 0.21.0-alpha.3
- Browser version<!-- (e.g. Chrome 95.0.4638.54) -->: Windows Node v18.6.0
- Any other relevant information:

<!-- If you are building Pyodide by yourself, please also include these information: -->

- Commit hash of Pyodide git repository: 843c024de35d5b15352eea8e4ae0a1b4eb6aec93
- Build environment<!--(e.g. Ubuntu 18.04, pyodide/pyodide-env:19 docker)- ->: Windows -> pyodide/pyodide-env:20 docker

### Additional context

<!-- Add any other context about the problem here. -->

--- a/docs/project/changelog.md
+++ b/docs/project/changelog.md
@@ -107,6 +107,9 @@ substitutions:
   empty container. Otherwise it returns `True`.
   {pr}`2803`
 
+- {{ Fix }} Fix `loadPyodide` errors for the Windows Node environment.
+  {pr}`2888`
+
 ### REPL
 
 - {{ Enhancement }} Add a spinner while the REPL is loading

--- a/src/js/compat.ts
+++ b/src/js/compat.ts
@@ -8,7 +8,7 @@ export const IN_NODE =
   typeof process.browser ===
     "undefined"; /* This last condition checks if we run the browser shim of process */
 
-let nodePathMod: any;
+let nodeUrlMod: any;
 let nodeFetch: any;
 let nodeVmMod: any;
 /** @private */
@@ -30,7 +30,7 @@ export async function initNodeModules() {
     return;
   }
   // @ts-ignore
-  nodePathMod = (await import("path")).default;
+  nodeUrlMod = (await import("url")).default;
   nodeFsPromisesMod = await import("fs/promises");
   if (globalThis.fetch) {
     nodeFetch = fetch;
@@ -194,6 +194,6 @@ async function nodeLoadScript(url: string) {
   } else {
     // Otherwise, hopefully it is a relative path we can load from the file
     // system.
-    await import(nodePathMod.resolve(url));
+    await import(nodeUrlMod.pathToFileURL(url).href);
   }
 }

--- a/src/js/pyodide.ts
+++ b/src/js/pyodide.ts
@@ -171,6 +171,9 @@ function calculateIndexURL(): string {
   const indexOfLastSlash = fileName.includes("/")
     ? fileName.lastIndexOf("/")
     : fileName.lastIndexOf("\\");
+  if (indexOfLastSlash === -1) {
+    throw new Error("Could not extract indexURL path from pyodide module location");
+  }
   return fileName.slice(0, indexOfLastSlash);
 }
 

--- a/src/js/pyodide.ts
+++ b/src/js/pyodide.ts
@@ -168,7 +168,8 @@ function calculateIndexURL(): string {
     err = e as Error;
   }
   let fileName = ErrorStackParser.parse(err)[0].fileName!;
-  return fileName.slice(0, fileName.lastIndexOf("/"));
+  const indexOfLastSlash = fileName.includes("/") ? fileName.lastIndexOf("/") : fileName.lastIndexOf("\\");
+  return fileName.slice(0, indexOfLastSlash);
 }
 
 /**

--- a/src/js/pyodide.ts
+++ b/src/js/pyodide.ts
@@ -172,7 +172,9 @@ function calculateIndexURL(): string {
     ? fileName.lastIndexOf("/")
     : fileName.lastIndexOf("\\");
   if (indexOfLastSlash === -1) {
-    throw new Error("Could not extract indexURL path from pyodide module location");
+    throw new Error(
+      "Could not extract indexURL path from pyodide module location"
+    );
   }
   return fileName.slice(0, indexOfLastSlash);
 }

--- a/src/js/pyodide.ts
+++ b/src/js/pyodide.ts
@@ -168,7 +168,9 @@ function calculateIndexURL(): string {
     err = e as Error;
   }
   let fileName = ErrorStackParser.parse(err)[0].fileName!;
-  const indexOfLastSlash = fileName.includes("/") ? fileName.lastIndexOf("/") : fileName.lastIndexOf("\\");
+  const indexOfLastSlash = fileName.includes("/")
+    ? fileName.lastIndexOf("/")
+    : fileName.lastIndexOf("\\");
   return fileName.slice(0, indexOfLastSlash);
 }
 

